Vungle/VungleAdapter/ALVungleMediationAdapter.m

```diff
@@ -877,3 +911,135 @@ - (void)vungleCreative:(nullable NSString *)creativeID readyForPlacement:(nullab
+        }];
+        [self.nativeAdDelegate didLoadAdForNativeAd: maNativeAd withExtraInfo: nil];
+    });
+}
+
+- (void)unregisterNativeAd
+{
+    if ( self.vungleNativeAd )
+    {
+        [self.vungleNativeAd unregisterView];
+    }
+}
+
+- (void)loadVungleNativeAd:(NSString *)placementIdentifier
+{
+    self.vungleNativeAd = [[VungleNativeAd alloc] initWithPlacementID: placementIdentifier];
+    self.vungleNativeAd.delegate = self;
+    self.vungleNativeAd.adOptionsPosition = NativeAdOptionsPositionTopRight;
+    [self.vungleNativeAd loadAdWithAdMarkup: self.parameters.bidResponse];
+}
+
+- (void)nativeAdDidLoad:(VungleNativeAd *)vungleNativeAd
+{
+    [self playNativeAd];
+}
+
+- (void)vungleNativeAd:(VungleNativeAd *)vungleNativeAd didFailWithError:(NSError *)error
+{
+    MAAdapterError *adapterError = [ALVungleMediationAdapter toMaxError: error];
+    [self.nativeAdDelegate didFailToLoadNativeAdWithError: adapterError];
+}
+
+- (void)nativeAdDidClick:(VungleNativeAd *)vungleNativeAd
+{
+    [self.nativeAdDelegate didClickNativeAd];
+}
+
+- (void)nativeAdDidTrackImpression:(VungleNativeAd *)vungleNativeAd
+{
+    [self.nativeAdDelegate didDisplayNativeAdWithExtraInfo: nil];
+}
+
+@end
+
+@implementation MAVungleNativeAd
+
+- (instancetype)initWithParentAdapter:(ALVungleMediationNativeAdAdapter *)parentAdapter builderBlock:(NS_NOESCAPE MANativeAdBuilderBlock)builderBlock
+{
+    self = [super initWithFormat: MAAdFormat.native builderBlock: builderBlock];
+    if ( self )
+    {
+        self.parentAdapter = parentAdapter;
+    }
+    return self;
+}
+
+- (void)prepareViewForInteraction:(MANativeAdView *)nativeAdView
+{
+    if ( !self.parentAdapter.vungleNativeAd )
+    {
+        return;
+    }
+    
+    NSMutableArray *clickableViews = [NSMutableArray array];
+    if ( [self.title al_isValidString] && nativeAdView.titleLabel )
+    {
+        [clickableViews addObject: nativeAdView.titleLabel];
+    }
+    if ( [self.body al_isValidString] && nativeAdView.bodyLabel )
+    {
+        [clickableViews addObject: nativeAdView.bodyLabel];
+    }
+    if ( [self.callToAction al_isValidString] && nativeAdView.callToActionButton )
+    {
+        [clickableViews addObject: nativeAdView.callToActionButton];
+    }
+    if ( self.icon && nativeAdView.iconImageView )
+    {
+        [clickableViews addObject: nativeAdView.iconImageView];
+    }
+    if ( self.mediaView && nativeAdView.mediaContentView )
+    {
+        [clickableViews addObject: nativeAdView.mediaContentView];
+    }
+    
+    [self.parentAdapter.vungleNativeAd registerViewForInteraction: nativeAdView
+                                                        mediaView: (VungleMediaView *) self.mediaView
+                                                    iconImageView: nativeAdView.iconImageView
+                                                   viewController: [ALUtils topViewControllerFromKeyWindow]
```

Comment:
We may want to keep an eye on the method you're using to get the presenting view controller here. Depending on how the method has been implemented, it may be using deprecated APIs.